src/chain/chain-generic-numerator.h

```diff
@@ -98,71 +96,79 @@ class GenericNumeratorComputation {
   bool Backward(CuMatrixBase<BaseFloat> *nnet_output_deriv);
 
  private:
-
   // Defining this constant as an enum is easier.  it controls a memory/speed
   // tradeoff, determining how many frames' worth of the transposed derivative
   // we store at a time.  It's not very critical; the only disadvantage from
   // setting it small is that we have to invoke an AddMat kernel more times.
   enum { kMaxDerivTimeSteps = 8 };
 
-  // sets up the alpha for frame t = 0.
-  void AlphaFirstFrame();
+  // for the remapped fsts, copy the appropriate activations to CPU memory
+  void CopySpecificPdfsIndirect(int sequence_id,
+                             const std::vector<MatrixIndexT> &indices,
+                             Matrix<BaseFloat> *output);
+  // for the remapped fsts, copy the computed values back to gpu,
+  // expand to the original shape and add to the output matrix
+  void AddSpecificPdfsIndirect(int sequence_id,
+                             const Matrix<BaseFloat> &logprobs,
+                             const std::vector<MatrixIndexT> &indices,
+                             CuMatrixBase<BaseFloat> *output);
 
-  // the alpha computation for some 0 < t <= num_time_steps_.
-  void AlphaGeneralFrame(int32 t);
+  // sets up the alpha for frame t = 0.
+  void AlphaFirstFrame(int seq);
 
-  BaseFloat ComputeTotLogLike();
+  // the alpha computation for 0 < t <= supervision_.frames_per_sequence
+  // for some 0 <= seq < supervision_.num_sequences.
+  BaseFloat AlphaGeneralFrame(int32 seq);
 
-  // sets up the beta for frame t = num_time_steps_.
-  void BetaLastFrame();
+  // the beta computation for 0 <= t < supervision_.frames_per_sequence
+  // for some 0 <= seq < supervision_.num_sequences.
+  void BetaLastFrame(int seq);
 
-  // the beta computation for 0 <= beta < num_time_steps_.
-  void BetaGeneralFrame(int32 t);
+  // the beta computation for t = supervision_.frames_per_sequence
```

Comment:
Are these comments swapped somehow?  Surely BetaGeneralFrame should not be for the last frame.
And I'm not sure it makes sense to have two functions?